File: intelligence/app/schemas/instamart.py
"""Swiggy Instamart (grocery) schemas.

Tool contracts verified against https://mcp.swiggy.com/builders/docs/reference/instamart/*.
search_products returns products[].variations[] with {spinId, quantity, price,
stock} — skuId isn't in the documented example but Swiggy's own DIY notebook
reads it off the same variation objects, so we parse it defensively (like the
rest of this codebase treats Swiggy's "loosely documented" envelopes) and fall
back to spinId if it's absent.
"""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field

from app.schemas.recipe import Ingredient

# Swiggy Instamart's own minimum cart value for checkout.
INSTAMART_MIN_CART_INR = 99
# Same beta safety cap already enforced on Food orders (ORDER_VALUE_CAP_INR) —
# applied here defensively since the docs flag ₹1000 as the typical ceiling too.
INSTAMART_ORDER_VALUE_CAP_INR = 1000


class InstamartVariation(BaseModel):
    spin_id: str
    sku_id: Optional[str] = None
    quantity: Optional[str] = None
    price: Optional[float] = None
    stock: Optional[bool] = None
    image_url: Optional[str] = None


class InstamartProduct(BaseModel):
    name: str
    brand: Optional[str] = None
    availability: bool = True
    variations: list[InstamartVariation] = Field(default_factory=list)


class MatchIngredientsRequest(BaseModel):
    items: list[Ingredient]
    address_id: str


class SearchProductsResponse(BaseModel):
    success: bool
    products: list[InstamartProduct] = Field(default_factory=list)
    error: Optional[str] = None


class MatchedIngredient(BaseModel):
    ingredient: Ingredient
    matched: bool
    confidence: float = 0.0
    product: Optional[InstamartProduct] = None
    variation: Optional[InstamartVariation] = None


class MatchIngredientsResponse(BaseModel):
    success: bool
    address_id: Optional[str] = None
    matches: list[MatchedIngredient] = Field(default_factory=list)
    error: Optional[str] = None
    address_required: bool = False


class InstamartCartItemInput(BaseModel):
    spin_id: str
    sku_id: Optional[str] = None
    quantity: int = 1


class UpdateInstamartCartRequest(BaseModel):
    address_id: str
    items: list[InstamartCartItemInput]


class InstamartCartLine(BaseModel):
    spin_id: str
    name: Optional[str] = None
    price: Optional[float] = None
    quantity: int = 1


class InstamartCartResponse(BaseModel):
    success: bool
    items: list[InstamartCartLine] = Field(default_factory=list)
    subtotal: Optional[float] = None
    total: Optional[float] = None
    available_payment_methods: list[str] = Field(default_factory=list)
    error: Optional[str] = None
    address_required: bool = False
    below_minimum: bool = False
    minimum_amount: int = INSTAMART_MIN_CART_INR


class InstamartCheckoutRequest(BaseModel):
    address_id: str
    payment_method: Optional[str] = None
    intent_app: Optional[str] = None
    generate_upi_qr: bool = False
    # Mandatory explicit user confirmation, same rule as Food's place_order.
    confirmed: bool = False


class InstamartCheckoutResponse(BaseModel):
    success: bool
    order_id: Optional[str] = None
    paas_id: Optional[str] = None
    status: Optional[str] = None
    pending_payment: bool = False
    message: Optional[str] = None
    error: Optional[str] = None
    address_required: bool = False
    cap_exceeded: bool = False
    cap_amount: int = INSTAMART_ORDER_VALUE_CAP_INR
    below_minimum: bool = False
    minimum_amount: int = INSTAMART_MIN_CART_INR


class PaymentOptionsResponse(BaseModel):
    success: bool
    methods: list[dict] = Field(default_factory=list)
    error: Optional[str] = None


class PaymentStatusResponse(BaseModel):
    success: bool
    status: Optional[str] = None
    order_id: Optional[str] = None
    error: Optional[str] = None


class TrackInstamartOrderResponse(BaseModel):
    success: bool
    order_id: Optional[str] = None
    status: Optional[str] = None
    eta: Optional[str] = None
    raw: Optional[dict] = None
    error: Optional[str] = None

File: mobile/src/services/instamart.ts
import { API_BASE_URL, getHeaders } from './apiBase';

export interface Ingredient {
  name: string;
  quantity: string;
  unit: string;
  notes?: string | null;
}

export interface InstamartVariation {
  spinId: string;
  skuId?: string | null;
  quantity?: string | null;
  price?: number | null;
  stock?: boolean | null;
  imageUrl?: string | null;
}

export interface InstamartProduct {
  name: string;
  brand?: string | null;
  availability: boolean;
  variations: InstamartVariation[];
}

export interface MatchedIngredient {
  ingredient: Ingredient;
  matched: boolean;
  confidence: number;
  product?: InstamartProduct | null;
  variation?: InstamartVariation | null;
}

export interface InstamartCartLine {
  spinId: string;
  name?: string | null;
  price?: number | null;
  quantity: number;
}

export interface InstamartCartState {
  success: boolean;
  items: InstamartCartLine[];
  subtotal?: number | null;
  total?: number | null;
  availablePaymentMethods: string[];
  error?: string;
  addressRequired?: boolean;
  belowMinimum?: boolean;
  minimumAmount: number;
}

export interface InstamartCheckoutResult {
  success: boolean;
  orderId?: string | null;
  paasId?: string | null;
  status?: string | null;
  pendingPayment: boolean;
  message?: string | null;
  error?: string;
  capExceeded?: boolean;
  capAmount?: number;
  belowMinimum?: boolean;
  minimumAmount?: number;
}

function mapVariation(v: Record<string, unknown> | undefined): InstamartVariation | null {
  if (!v) return null;
  return {
    spinId: v.spin_id as string,
    skuId: v.sku_id as string | null,
    quantity: v.quantity as string | null,
    price: v.price as number | null,
    stock: v.stock as boolean | null,
    imageUrl: v.image_url as string | null,
  };
}

function mapProduct(p: Record<string, unknown>): InstamartProduct {
  return {
    name: p.name as string,
    brand: p.brand as string | null,
    availability: !!p.availability,
    variations: ((p.variations as Record<string, unknown>[]) || [])
      .map(mapVariation)
      .filter((v): v is InstamartVariation => v !== null),
  };
}

export async function searchInstamartProducts(
  query: string,
  addressId: string,
): Promise<{ success: boolean; products: InstamartProduct[]; error?: string }> {
  const params = new URLSearchParams({ query, address_id: addressId });
  const res = await fetch(`${API_BASE_URL}/instamart/search-products?${params.toString()}`, {
    headers: await getHeaders(),
  });
  const data = await res.json();
  return {
    success: !!data.success,
    products: ((data.products as Record<string, unknown>[]) || []).map(mapProduct),
    error: data.error,
  };
}

function mapMatch(m: Record<string, unknown>): MatchedIngredient {
  return {
    ingredient: m.ingredient as Ingredient,
    matched: !!m.matched,
    confidence: (m.confidence as number) ?? 0,
    product: m.product ? mapProduct(m.product as Record<string, unknown>) : null,
    variation: mapVariation(m.variation as Record<string, unknown> | undefined),
  };
}

export async function matchIngredients(
  items: Ingredient[],
  addressId: string,
): Promise<{ success: boolean; matches: MatchedIngredient[]; error?: string }> {
  const res = await fetch(`${API_BASE_URL}/instamart/match-ingredients`, {
    method: 'POST',
    headers: await getHeaders(),
    body: JSON.stringify({ items, address_id: addressId }),
  });
  const data = await res.json();
  return {
    success: !!data.success,
    matches: ((data.matches as Record<string, unknown>[]) || []).map(mapMatch),
    error: data.error,
  };
}

function mapCart(data: Record<string, unknown>): InstamartCartState {
  return {
    success: !!data.success,
    items: (data.items as InstamartCartLine[]) || [],
    subtotal: data.subtotal as number | null,
    total: data.total as number | null,
    availablePaymentMethods: (data.available_payment_methods as string[]) || [],
    error: data.error as string | undefined,
    addressRequired: !!data.address_required,
    belowMinimum: !!data.below_minimum,
    minimumAmount: (data.minimum_amount as number) ?? 99,
  };
}

export async function getInstamartCart(addressId: string): Promise<InstamartCartState> {
  const res = await fetch(`${API_BASE_URL}/instamart/cart?address_id=${encodeURIComponent(addressId)}`, {
    headers: await getHeaders(),
  });
  const data = await res.json();
  return mapCart(data);
}

/** Always sends the full desired cart state — update_cart replaces, not appends. */
export async function updateInstamartCart(
  addressId: string,
  items: { spinId: string; skuId?: string | null; quantity: number }[],
): Promise<InstamartCartState> {
  const res = await fetch(`${API_BASE_URL}/instamart/cart`, {
    method: 'POST',
    headers: await getHeaders(),
    body: JSON.stringify({
      address_id: addressId,
      items: items.map((i) => ({ spin_id: i.spinId, sku_id: i.skuId, quantity: i.quantity })),
    }),
  });
  const data = await res.json();
  return mapCart(data);
}

export async function checkoutInstamart(
  addressId: string,
  paymentMethod: string | undefined,
  confirmed: boolean,
): Promise<InstamartCheckoutResult> {
  const res = await fetch(`${API_BASE_URL}/instamart/checkout`, {
    method: 'POST',
    headers: await getHeaders(),
    body: JSON.stringify({ address_id: addressId, payment_method: paymentMethod, confirmed }),
  });
  const data = await res.json();
  return {
    success: !!data.success,
    orderId: data.order_id,
    paasId: data.paas_id,
    status: data.status,
    pendingPayment: !!data.pending_payment,
    message: data.message,
    error: data.error,
    capExceeded: !!data.cap_exceeded,
    capAmount: data.cap_amount,
    belowMinimum: !!data.below_minimum,
    minimumAmount: data.minimum_amount,
  };
}
